mesh/mesh.go

```diff
@@ -194,68 +146,71 @@ func (m *Mesh) ValidateLayer(lyr *types.Layer) {
 	m.Info("done validating layer %v", lyr.Index())
 }
 
-func SortBlocks(blocks []*types.Block) []*types.Block {
-	//not final sorting method, need to talk about this
-	sort.Slice(blocks, func(i, j int) bool { return blocks[i].ID() < blocks[j].ID() })
-	return blocks
-}
-
-func (m *Mesh) ExtractUniqueOrderedTransactions(l *types.Layer) []*Transaction {
-	txs := make([]*Transaction, 0, layerSize)
-	sortedBlocks := SortBlocks(l.Blocks())
-
-	txids := make(map[types.TransactionId]struct{})
+func (m *Mesh) ExtractUniqueTransactions(l *types.Layer) (validBlocks, invalidBlocks []*types.Transaction) {
+	txids := map[bool]map[types.TransactionId]struct{}{
+		true:  make(map[types.TransactionId]struct{}),
+		false: make(map[types.TransactionId]struct{}),
+	}
 
-	for _, b := range sortedBlocks {
+	for _, b := range l.Blocks() {
 		valid, err := m.ContextualValidity(b.ID())
+		if err != nil {
+			m.With().Error("could not get contextual validity for block", log.BlockId(uint64(b.ID())), log.Err(err))
+		}
 		events.Publish(events.ValidBlock{Id: uint64(b.ID()), Valid: valid})
 		if !valid {
-			if err != nil {
-				m.With().Error("could not get contextual validity for block", log.BlockId(uint64(b.ID())), log.Err(err))
-			}
-			m.With().Info("block not contextually valid", log.BlockId(uint64(b.ID()))) // TODO: do we want this log? is it info?
-			continue
+			m.With().Warning("block not contextually valid", log.BlockId(uint64(b.ID()))) // TODO: do we want this log?
```

Comment:
I just skimmed the tortoise code and couldn't find another place this is logged. I agree that this isn't a great place to log it, though.

Should I move this log (and the event publication, above) to where contextual validity is determined?

Reply:
Specifically, I'd add to following code:
https://github.com/spacemeshos/go-spacemesh/blob/122bfa3dc1fb83ac3754a200f8393944ac0f3a20/tortoise/ninja_tortoise.go#L555-L560
this logic:
```go
if vote == Against {
	ni.With().Warning("block not contextually valid", log.BlockId(uint64(bid)))
}
events.Publish(events.ValidBlock{Id: uint64(bid), Valid: vote == Support})
```

Reply:
1. It is not clear that we need this log. For debugging purposes it is good.
2. The right place to enter this log, in my opinion, is under SaveOpinion(-;) you wouldn't need the if, only else.
3. I'd write "block is contextually invalid"

Reply:
what barak said, also better to put events in algorithm.go  not here

Reply:
1. You mean "not clear that we need this log" **here**, or **anywhere**?
2. I actually don't need the `if` here, because there's already a check for `Abstain` and then for `Support`. I don't have a good understanding of this area - when `SaveOpinion` is called there are no longer any `Abstain` votes? Otherwise, I **would** need an `if` there (not that it's important).
3. Yeah, I agree.

The reason I originally wanted to put it here and not in `SaveOpinion` is that an opinion about a block can be saved more than once (if the block is part of more than one voting patterns). Now I think that the same applies to where I suggested to add the logic. Am I right? Does it even matter? (the log appearing more than once isn't a big deal, but the event -- not sure)

Reply:
1. anywhere/at all.
2. Correct, save opinion is either true/(contextually) valid or false/invalid.

If I understand you correctly, you are not right. SaveOpinion() should happen only once, after we have a firm opinion (thus not Abstain) and it is irreversible - so basically we call this method only for blocks below p_base.